### PR TITLE
feat(): Angularfire auth guards

### PR DESCRIPTION
**Pull request by jamesdaniels**
### Checklist

   - Issue number for this PR: #nnn (required)
   - Docs included?: (yes/no; required for all API/functional changes) 
   - Test units included?: (yes/no; required) 
   - In a clean directory, `yarn install`, `yarn test` run successfully? (yes/no; required)

### Description

Allow developers to protect routes in their Angular application using Firebase authentication.

#### TODOs

- [ ] Get test coverage
- [ ] Polish the documentation

#### Basic example

```ts
import { AngularFireAuthGuard } from '@angular/fire/auth-guard';

export const routes: Routes = [
    { path: '',      component: AppComponent },
    { path: 'items', component: ItemListComponent, canActivate: [AngularFireAuthGuard] },
]
```

#### Use our pre-built pipes for common tests

```ts
import { AngularFireAuthGuard, hasCustomClaim, redirectUnauthorizedTo, redirectLoggedInTo } from '@angular/fire/auth-guard';

const adminOnly = hasCustomClaim('admin');
const redirectUnauthorizedToLogin = redirectUnauthorizedTo(['login']);
const redirectLoggedInToItems = redirectLoggedInTo(['items']);
const belongsToAccount = (next) => hasCustomClaim(`account-${next.params.id}`);

export const routes: Routes = [
    { path: '',      component: AppComponent },
    { path: 'login', component: LoginComponent,        canActivate: [AngularFireAuthGuard], data: { authGuardPipe: redirectLoggedInToItems }},
    { path: 'items', component: ItemListComponent,     canActivate: [AngularFireAuthGuard], data: { authGuardPipe: redirectUnauthorizedToLogin },
    { path: 'admin', component: AdminComponent,        canActivate: [AngularFireAuthGuard], data: { authGuardPipe: adminOnly }},
    { path: 'accounts/:id', component: AdminComponent, canActivate: [AngularFireAuthGuard], data: { authGuardPipe: belongsToAccount }}
];
```

#### Increase readability with our `canActivate` helper

```ts
import { canActivate } from '@angular/fire/auth-guard';

export const routes: Routes = [
    { path: '',             component: AppComponent },
    { path: 'login',        component: LoginComponent,    ...canActivate(redirectLoggedInToItems) },
    { path: 'items',        component: ItemListComponent, ...canActivate(redirectUnauthorizedToLogin) },
    { path: 'admin',        component: AdminComponent,    ...canActivate(adminOnly) },
    { path: 'accounts/:id', component: AdminComponent,    ...canActivate(belongsToAccount) }
];
```

#### Compose your own pipes

```ts
import { pipe, of } from 'rxjs';
import { map, switchMap } from 'rxjs/operators';
import { customClaims } from '@angular/fire/auth-guard';

const editorOnly = pipe(customClaims, map(claims => claims.role === "editor"));
const redirectToProfileEditOrLogin = map(user => user ? ['profiles', user.uid, 'edit'] : ['login']);
const onlyAllowSelf = (next) => map(user => !!user && next.params.userId === user.uid);
const accountAdmin = (next) => pipe(customClaims, map(claims => claims[`account-${next.params.accountId}-role`] === "admin"));
```

### CODE REVIEW

**Comment by jamesdaniels** [comment]
@davideast feel free to start poking at this. Going to work on the test coverage and documentation in the meantime. Targeting 5.2 release.

**Comment by jamesdaniels** [comment]
FYI this has been released for testing on `@next` as of `5.2.0-beta.3`

**Comment by hiepxanh** [comment]
wow, so convinent :D thank you so much

**Comment by jamesdaniels** [comment]
Assigning @davideast for API review. @jhuleatt if you could give a quick code review.

**Comment by jamesdaniels** [comment]
FYI tests and docs need more work.

File: docs/auth/router-guards.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+```ts
+import { AngularFireAuthGuard } from '@angular/fire/auth-guard';
+
+export const routes: Routes = [
+    { path: '',      component: AppComponent },
+    { path: 'items', component: ItemListComponent, canActivate: [AngularFireAuthGuard] },
+]
+```
+
+## Use our pre-built pipes for common tests
```
[the remark this answers is not in the document]
Comment (jhuleatt) [reply docs/auth/router-guards.md]: Also, it looks like there are a lot of really useful pre-built pipes in auth-guard.ts. Maybe just list them all in this section?

File: src/auth-guard/auth-guard.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+}
+
+export const canActivate = (pipe: AuthPipe|AuthPipeGenerator) => ({
+    canActivate: [ AngularFireAuthGuard ], data: { authGuardPipe: pipe.name === "" ? pipe : () => pipe}
+});
+
+export const loggedIn: AuthPipe = map(user => !!user);
+export const isNotAnonymous: AuthPipe = map(user => !!user && !user.isAnonymous);
+export const idTokenResult = switchMap((user: User|null) => user ? user.getIdTokenResult() : of(null));
```
Comment (jhuleatt) [inline src/auth-guard/auth-guard.ts]: I think `getIdTokenResult` returns a Promise - can Angular handle that?

File: src/auth-guard/auth-guard.spec.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        })();
+    });
+
+    afterEach(done => {
+        app.delete().then(done, done.fail);
+    });
+    
+    it('should be injectable', () => {
+        expect(router).toBeTruthy();
+    });
```
Comment (jhuleatt) [inline src/auth-guard/auth-guard.spec.ts]: [here's a StackOverflow answer](https://stackoverflow.com/a/53881037/4816918) with suggestions on ways to test auth guards. But maybe that's overkill and we just have a dummy user object and quick happy path tests for the different built-in pipes for now?

File: docs/auth/router-guards.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
+# Route users with AngularFire guards
+
```
Comment (jhuleatt) [inline docs/auth/router-guards.md]: For a "what this solves" blurb at the top, maybe something like "These are convenience pipes provided by AngularFire to protect routes using Angular's [Route Guards](https://angular.io/guide/router#milestone-5-route-guards)."

File: src/auth-guard/auth-guard.spec.ts (hunk 3 above)
Comment (jhuleatt) [reply src/auth-guard/auth-guard.spec.ts]: Maybe also spy on the AngularFireAuthGuard pipe to make sure it calls the pipe provided to it with the data provided to it?

File: src/auth-guard/auth-guard.ts (hunk 2 above)
Comment (jamesdaniels) [reply src/auth-guard/auth-guard.ts]: Yeah, switchMap will handle that automatically

File: src/auth-guard/auth-guard.spec.ts (hunk 3 above)
Comment (jamesdaniels) [reply src/auth-guard/auth-guard.spec.ts]: I think I'll have to put together a proper E2E for this, something I'm working on on another branch. Think I'll punt on building more robust tests for now & take it as an action item.

**Comment by svzi** [comment]
Any idea when there will be a new release with the merged PR?

**Comment by jamesdaniels** [comment]
@svzi you can give this a try now with the beta 4 release. `npm i --save @angular/fire@next`.

I'm hoping to release 5.2 stable early next week.

**Comment by svzi** [comment]
@jamesdaniels Thanks a lot! But it seems like I'm missing something, or there is something not working right now:

```
ERROR in node_modules/@angular/fire/auth-guard/auth-guard.d.ts(11,5): error TS2416: Property 'canActivate' in type 'AngularFireAuthGuard' is not assignable to the same property in base type 'CanActivate'.
  Type '(next: ActivatedRouteSnapshot, state: RouterStateSnapshot) => Observable<boolean | UrlTree>' is not assignable to type '(route: ActivatedRouteSnapshot, state: RouterStateSnapshot) => boolean | UrlTree | Observable<boolean | UrlTree> | Promise<boolean | UrlTree>'.
    Type 'Observable<boolean | UrlTree>' is not assignable to type 'boolean | UrlTree | Observable<boolean | UrlTree> | Promise<boolean | UrlTree>'.
```

**Comment by jamesdaniels** [comment]
@svzi which version of typescript / Ng?

**Comment by svzi** [comment]
@jamesdaniels Angular 7.2.0 (CLI 7.3.9) and TypeScript 3.2.2.

**Comment by jamesdaniels** [comment]
I'll look into this. Will track here #2086

**Comment by svzi** [comment]
Thanks a lot, appreciate your support!

**Pull request merged**